Added cumulative metrics frontend to pod list.

PiotrDabkowski opened the pull request:
Added first version of graph to the pod list page. 
![nnn](https://cloud.githubusercontent.com/assets/4052440/18050482/15ef494e-6def-11e6-92a0-3188ea7303b7.png)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/dashboard/1142)
<!-- Reviewable:end -->

bryk commented:
@olekzabl Can you review this? (when Piotrek is back)

bryk commented:
Please rebase and publish new changes. 

---

Reviewed 3 of 16 files at r1.
Review status: 3 of 16 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ-jBwULKPApshPPwP1:b-necvww)*
<!-- Sent from Reviewable.io -->

bryk commented:
A few comments, PTAL 

---

Reviewed 7 of 16 files at r1, 1 of 6 files at r2.
Review status: 8 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[bower.json, line 37 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ0-oJpWz6_JpYIAuvq:-KQ0-oJpWz6_JpYIAuvr:bwn31ce) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/bower.json#L37)):*
> ```JSON
>   },
>   "devDependencies": {
>     "cljsjs-packages-externs": "https://github.com/cljsjs/packages.git",
> ```

Can you pin this dependency to a specific SHA version? It is bad to have unversioned deps. 

---

*[src/app/frontend/common/components/graph/graph.html, line 17 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ007i8MfOzmVVKLQ3F:-KQ007i8MfOzmVVKLQ3G:b-62sykv) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph.html#L17)):*
> ```HTML
> -->
> 
> <div id="metricChart">
> ```

You shouldn't be using IDs in components. This makes it impossible to add two graphs to a page. The component should be self-contained and isolated. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 104 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01152rEpubBRFnXLA:-KQ01153c6FWrmwYxQv0:b-ybzj2i) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L104)):*
> ```JavaScript
>       ];
> 
>       let graphArea = d3.select('#metricChart');
> ```

Instead of ids, you should select this directive's element, by, like:

```js
          let  chart = d3.select($element[0]);
```

`$element` can be injected to the constructor. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 108 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01Fg1UXpM8B2g7eRP:-KQ01Fg2cq8KTaYs1FiD:b5amcxj) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L108)):*
> ```JavaScript
> 
>       svg.attr('height', 300).datum(data).call(chart);
>       svg.style({
> ```

Can you set CSS class to this SVG and style all in the CSS? This it to have all styling code (like colors) in CSSes. Should work fine. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 126 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01VK_j-iZPFh28TK0:-KQ01VK_j-iZPFh28TK1:b-iqy3dr) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L126)):*
> ```JavaScript
>       // );
> 
>       // todo remove this block after enabling intelligent graph resize above.
> ```

Hmm... Can you remove this and make the graph unresizable for now? And then later implement the "intelligent" solution :) 

We don't need to have this temp solution here. 

---

*[src/app/frontend/common/components/sparkline/sparkline.html, line 21 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01rTfJrNANVNUP53c:-KQ01rTfJrNANVNUP53d:b-purgya) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/sparkline/sparkline.html#L21)):*
> ```HTML
>      class="kd-sparkline">
>   <polygon ng-attr-points="0,1 {{::sparklineCtrl.polygonPoints()}} 1,1"
>   ng-class=""/>
> ```

Hmm.. Why this change? Can you remove ng-class entirely? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ0228JzK99FOdPecfN:bnsj53x)*
<!-- Sent from Reviewable.io -->

bryk commented:
A few more comments about tests and styles, PTAL.

---

Review status: 8 of 13 files reviewed at latest revision, 17 unresolved discussions.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 18 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ038GmfasbR3w_lunU:-KQ038GmfasbR3w_lunV:bgaqi4n) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L18)):*
> ```JavaScript
> import memoryFilter from 'common/filters/memory_filter';
> 
> function precisionFilter(d) {
> ```

You can easily test this function. Just export it and cover all branches. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 29 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ031WkU7XrJPqZfK0z:-KQ031WkU7XrJPqZfK1-:b-g3tsu9) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L29)):*
> ```JavaScript
> class GraphController {
>   constructor() {
>     /** @export {!Array<!backendApi.Metric>} */
> ```

Add a comment that this is initialized from a binding. See how other components do this. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 33 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ02CK9VIr29chMyXkV:-KQ02CK9VIr29chMyXkW:b-6w1ggk) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L33)):*
> ```JavaScript
> 
>     // check if metrics are available
>     if (this.metrics instanceof Array && this.metrics.length !== 0) {
> ```

Please move all initialization logic to `$onInit` method. This makes testing easy, as you have no logic insde your controller and you can easily instantiate things and mock them. 

`$onInit` is called just after constructor by angular. Reading: https://toddmotto.com/on-init-require-object-syntax-angular-component/

---

*[src/app/frontend/common/components/graph/graph_module.js, line 33 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ02pUNdwoFNLO5F7Y5:-KQ02pUNdwoFNLO5F7Y6:b-unwnjz) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L33)):*
> ```JavaScript
> 
>     // check if metrics are available
>     if (this.metrics instanceof Array && this.metrics.length !== 0) {
> ```

type of `this.metrics` says that it is always non-null array. Hence you should not be doing this safe check? 

You can do this check and throw an error if it is not fulfilled. Just to be safe. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 36 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03R4cV_9hX1_c61Fb:-KQ03R4cV_9hX1_c61Fc:b-hk4c61) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L36)):*
> ```JavaScript
>       // extract data points by metric name and render graph.
>       let toDraw = {};
>       for (let i = 0; i < this.metrics.length; i++) {
> ```

This piece of code can be easily tested. Just extract it and test whether `toDraw` is correct. 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 58 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03GVvYZX4pHs0dpMB:-KQ03GVvYZX4pHs0dpMC:b-hk8kfq) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L58)):*
> ```JavaScript
>       // chart sub-models (ie. xAxis, yAxis, etc) when accessed directly, return themselves, not the
>       // parent chart, so need to chain separately
>       chart.xAxis.axisLabel('Time')
> ```

Can you internationalize this? See how other components do i18n :) 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 61 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03kbWv2Chr3aHHrbD:-KQ03kbWv2Chr3aHHrbE:bf9z6qa) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L61)):*
> ```JavaScript
>           .tickFormat((d) => d3.time.format('%H:%M')(new Date(d)))
>           .staggerLabels(true);
>       chart.yAxis1.axisLabel('CPU (cores)').tickFormat(function(d) {
> ```

You can extract this function and test it independently.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 69 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03oU11oSfLN79Uqmf:-KQ03oU11oSfLN79Uqmg:bf9z6qa) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L69)):*
> ```JavaScript
>       });
> 
>       chart.yAxis2.axisLabel('Memory (bytes)').tickFormat(function(d) {
> ```

You can extract this function and test it independently.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 77 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03pCygNJDlu2elBln:-KQ03pCygNJDlu2elBlo:bf9z6qa) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L77)):*
> ```JavaScript
>       });
> 
>       chart.interactiveLayer.tooltip.valueFormatter(function(d, i) {
> ```

You can extract this function and test it independently.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 86 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03ryRWu2vyo1BebpP:-KQ03ryRWu2vyo1BebpQ:b2shqwa) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L86)):*
> ```JavaScript
>           values: parsedGraphData['cpu/usage_rate'],
>           key: 'Cpu Usage',
>           color: '#00c752',
> ```

Can you use CSS classes instead of hardcoded colors? 

---

*[src/app/frontend/common/components/graph/graph_module.js, line 146 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03awtYJ6392Bj4-tm:-KQ03awtYJ6392Bj4-tn:bkb922c) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L146)):*
> ```JavaScript
>   }
> }
> ```

Can you add `@type` info here? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ042Ei7mn0jP26cZPi:blezkbl)*
<!-- Sent from Reviewable.io -->

PiotrDabkowski commented:
@bryk Done everything except of styles, PTAL.

---

Review status: 7 of 17 files reviewed at latest revision, 18 unresolved discussions.

---

*[bower.json, line 37 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ0-oJpWz6_JpYIAuvq:-KQ082ar72VHvrejuBIT:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/bower.json#L37)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you pin this dependency to a specific SHA version? It is bad to have unversioned deps. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph.html, line 17 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ007i8MfOzmVVKLQ3F:-KQ083T30490RA7NFkRU:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph.html#L17)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You shouldn't be using IDs in components. This makes it impossible to add two graphs to a page. The component should be self-contained and isolated. 
</details>
Done.

---

*[src/app/frontend/common/components/sparkline/sparkline.html, line 21 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01rTfJrNANVNUP53c:-KQ0GFpIRj5rUFBkNv1A:bnvtzk3) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/sparkline/sparkline.html#L21)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Hmm.. Why this change? Can you remove ng-class entirely? 
</details>
I changed it in order to have 2 separate classes for sparklines - one for CPU and another for Memory. Removed unnecessary ng-class.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 18 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ038GmfasbR3w_lunU:-KQ08807R_FsDjmBtgTE:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L18)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You can easily test this function. Just export it and cover all branches. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 29 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ031WkU7XrJPqZfK0z:-KQ089nCUCZitGufkQpM:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L29)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Add a comment that this is initialized from a binding. See how other components do this. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 33 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ02CK9VIr29chMyXkV:-KQ0Bcits-tBMLO-Jr77:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L33)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Please move all initialization logic to `$onInit` method. This makes testing easy, as you have no logic insde your controller and you can easily instantiate things and mock them. 
> 
> `$onInit` is called just after constructor by angular. Reading: https://toddmotto.com/on-init-require-object-syntax-angular-component/
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 33 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ02pUNdwoFNLO5F7Y5:-KQ0BlwFED1AMBlvLHId:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L33)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> type of `this.metrics` says that it is always non-null array. Hence you should not be doing this safe check? 
> 
> You can do this check and throw an error if it is not fulfilled. Just to be safe. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 36 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03R4cV_9hX1_c61Fb:-KQ0xMw5JhHvRJeQA3Nl:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L36)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> This piece of code can be easily tested. Just extract it and test whether `toDraw` is correct. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 58 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03GVvYZX4pHs0dpMB:-KQ0cTenZ8EN6gfa97w7:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L58)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you internationalize this? See how other components do i18n :) 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 61 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03kbWv2Chr3aHHrbD:-KQ0xNdABoChV355sh3s:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L61)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You can extract this function and test it independently.
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 69 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03oU11oSfLN79Uqmf:-KQ0xOIdw5yo26_wlseV:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L69)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You can extract this function and test it independently.
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 77 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03pCygNJDlu2elBln:-KQ0xPLppvE1wc7_FWcd:b-biw8cw) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L77)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You can extract this function and test it independently.
</details>
This one just selects appropriate value formatter.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 86 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03ryRWu2vyo1BebpP:-KQ0cYxocJIoz-pDu4MH:by21fp4) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L86)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you use CSS classes instead of hardcoded colors? 
</details>
Using CSS complicates everything because I would have to define CSS selectors not only for lines but also for legend and tooltip. On the other hand specifying required styles as arguments to the nvd3 handles everything automatically.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 104 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01152rEpubBRFnXLA:-KQ0C2szi0bFwxsbexAR:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L104)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Instead of ids, you should select this directive's element, by, like:
> 
> ```js
>           let  chart = d3.select($element[0]);
> ```
> 
> `$element` can be injected to the constructor. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 108 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01Fg1UXpM8B2g7eRP:-KQ0G0IHZHoU8sE4lHmP:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L108)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you set CSS class to this SVG and style all in the CSS? This it to have all styling code (like colors) in CSSes. Should work fine. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 126 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ01VK_j-iZPFh28TK0:-KQ0Eg337hqKwfbmeJca:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L126)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Hmm... Can you remove this and make the graph unresizable for now? And then later implement the "intelligent" solution :) 
> 
> We don't need to have this temp solution here. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_module.js, line 146 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ03awtYJ6392Bj4-tm:-KQ0dQikVAa_hjF4-59K:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/6d40a1f3cfa4b4aeef6a32ecb39d620de18f6466/src/app/frontend/common/components/graph/graph_module.js#L146)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you add `@type` info here? 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ0dUU1InxCpSpvO1M-:b-b353xn)*
<!-- Sent from Reviewable.io -->

bryk commented:
PTAL

---

Reviewed 6 of 10 files at r3.
Review status: 13 of 17 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 68 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4X9e4cdIoRmW3p2fZ:-KQ4X9e4cdIoRmW3p2f_:b-cfu25f) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L68)):*
> ```JavaScript
> }
> 
> class GraphController {
> ```

This controller is not tested at all. Let's fix this :) 

In tests you can instantiate it by:
`let ctrl = $componentController('kdGraph', {}, {metrics: ...});`

Then you should initialize it:
`ctrl.$onInit`. 

---

*[src/app/frontend/common/components/graph/graph_component.js, line 95 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XQsbLb9JvobYCM1F:-KQ4XQsccBfFFE8ng2Id:bz94yml) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L95)):*
> ```JavaScript
>     }
>   }
> ```

We need `@type` information here. 

---

*[src/app/frontend/common/components/graph/graph_component.js, line 96 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4X_R8JpUOuZx9hZlo:-KQ4X_R8JpUOuZx9hZlp:bjzx0kh) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L96)):*
> ```JavaScript
>   }
> 
>   generateGraph(parsedGraphData) {
> ```

Should this be private? 

---

*[src/app/frontend/common/components/graph/graph_component.js, line 100 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4ZEHk5b73Ff_KvQQL:-KQ4ZEHk5b73Ff_KvQQM:bsiujon) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L100)):*
> ```JavaScript
>     let data;
> 
>     nv.addGraph(() => {
> ```

You can test this by executing the function and checking its HTML output. This should be easy. 

---

*[src/app/frontend/common/components/graph/graph_component.js, line 159 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4Z8pgSNB9IzHUGg37:-KQ4Z8ph7z-7OAZvZhDN:b-60o0c2) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L159)):*
> ```JavaScript
>       // update the graph in case of graph area resize
>       nv.utils.windowResize(chart.update);
>       this.scope_.$watch(
> ```

Let me think about other solutions to this....

---

*[src/app/frontend/common/components/graph/graph_component.js, line 161 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XpaD_cRva3eQJUY_:-KQ4XpaD_cRva3eQJUYa:bjdv9t6) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L161)):*
> ```JavaScript
>       this.scope_.$watch(
>           () => graphArea.node().getBoundingClientRect().width,  // variable to watch
>           () => setTimeout(
> ```

use angular service for setTimeout: `$window.setTimeout`

---

*[src/app/frontend/common/components/graph/graph_component.js, line 162 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XujbQFTo7xbLG45j:-KQ4Xujcg4CSkK5Nq85Z:bru6ji9) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L162)):*
> ```JavaScript
>           () => graphArea.node().getBoundingClientRect().width,  // variable to watch
>           () => setTimeout(
>               chart.update, 500),  // listener, call after 500ms after animation is complete.
> ```

Can this be 0? 

---

*[src/app/frontend/common/components/graph/graph_component.js, line 163 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XmlrSTaHJyJ0Ca8V:-KQ4XmlrSTaHJyJ0Ca8W:byh0j6h) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L163)):*
> ```JavaScript
>           () => setTimeout(
>               chart.update, 500),  // listener, call after 500ms after animation is complete.
>           true                     // deep watch
> ```

Why deep watch? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ4ZM-6QESXmJioXPSb:bb71pro)*
<!-- Sent from Reviewable.io -->

bryk commented:
Re this:
![Selection\_115.png](https://files.reviewableusercontent.io/images/reviewable/1159999/tJAdGerbJM3v7ORDbcLhcSRn/Selection_115.png)

1. Labels on X axis are wrong
2. Can you make the same number of ticks for left and right scale?
3.Can you make the scale from 0 to MAX? Let's talk offline about this. 

---

Review status: 13 of 17 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ4ghquX4i9fsCtYsuz:bdusrar)*
<!-- Sent from Reviewable.io -->

olekzabl commented:
Reviewed 7 of 16 files at r1, 1 of 6 files at r2, 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 19 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ5-tUahGHx5y-fpGb5:-KQ5-tUahGHx5y-fpGb6:b-atpvnw) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L19)):*
> ```JavaScript
> 
> /**
>  * Formats the number to contain ideally 3 significant figures, but at most 3 decimal places.
> ```

Is this true? How many decimal places will the function return for 0.00001?

---

*[src/app/frontend/common/components/graph/graph_component.js, line 23 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ54XdpdqzU6i3-Z2rt:-KQ54XdpdqzU6i3-Z2ru:bctp71x) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L23)):*
> ```JavaScript
>  * @return {string}
>  */
> function precisionFilter(d) {
> ```

private? (or tested? :)

---

*[src/app/frontend/common/components/graph/graph_component.js, line 61 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ50WmWa3RTnnsCYfC_:-KQ50WmXAq3hHT6sMLIi:bnycp7p) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L61)):*
> ```JavaScript
>   // extract data points by metric name
>   let dataPoints = {};
>   for (let i = 0; i < metrics.length; i++) {
> ```

I'd simplify this loop to "for (let metric of metrics) {...}"
(Note "of", not "in")
But I'm not sure it will work for you; you're using a different JavaScript than me :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

olekzabl commented:
Reviewed 2 of 6 files at r2.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

bryk commented:
@PiotrDabkowski Let us know when this is ready to review :)

PiotrDabkowski commented:
@bryk PTAL

---

Review status: 8 of 21 files reviewed at latest revision, 27 unresolved discussions.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 19 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ5-tUahGHx5y-fpGb5:-KQ5y20HcoFO9LiLDd1Q:b-92vuo2) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L19)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> Is this true? How many decimal places will the function return for 0.00001?
</details>
Good point, changed the doc. In practice the argument to the function will never be smaller than 0.001.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 23 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ54XdpdqzU6i3-Z2rt:-KQ5xJC-BQI10M0IZmqH:b1u6t23) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L23)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> private? (or tested? :)
</details>
private :) It is tested as part of formatMemoryUsage and formatCpuUsage

---

*[src/app/frontend/common/components/graph/graph_component.js, line 61 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ50WmWa3RTnnsCYfC_:-KQ5yL-xV6VukLmTlrz9:bnleqfy) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L61)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> I'd simplify this loop to "for (let metric of metrics) {...}"
> (Note "of", not "in")
> But I'm not sure it will work for you; you're using a different JavaScript than me :)
</details>
Thank you for this! Amazing keyword, I will use it everywhere now :)

---

*[src/app/frontend/common/components/graph/graph_component.js, line 68 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4X9e4cdIoRmW3p2fZ:-KQ575xo_E8IC4q_4WPI:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L68)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> This controller is not tested at all. Let's fix this :) 
> 
> In tests you can instantiate it by:
> `let ctrl = $componentController('kdGraph', {}, {metrics: ...});`
> 
> Then you should initialize it:
> `ctrl.$onInit`. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 95 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XQsbLb9JvobYCM1F:-KQ4t6mOtAmgQ85pOvlX:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L95)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> We need `@type` information here. 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 96 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4X_R8JpUOuZx9hZlo:-KQ4sNX2w3_3MVzPIuRU:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L96)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Should this be private? 
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 100 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4ZEHk5b73Ff_KvQQL:-KQ56wqfJFVRv8Ch-kyY:b53ihhm) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L100)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You can test this by executing the function and checking its HTML output. This should be easy. 
</details>
Added HTML test to onInit instead. 

---

*[src/app/frontend/common/components/graph/graph_component.js, line 161 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XpaD_cRva3eQJUY_:-KQ4rD30BGAd_o1TOGQq:b-fisbj1) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L161)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> use angular service for setTimeout: `$window.setTimeout`
</details>
setTimeout will soon be completely removed after different method of left hand side nav animation is implemented (instant DOM change)

---

*[src/app/frontend/common/components/graph/graph_component.js, line 162 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XujbQFTo7xbLG45j:-KQ4r50wYknpvDquuOA2:bribihm) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L162)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can this be 0? 
</details>
As above.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 163 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ4XmlrSTaHJyJ0Ca8V:-KQ4s28k8wpSyLTlxQ-L:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L163)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Why deep watch? 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQ4sRDAzIO2xaMfceQA:by7e27m)*
<!-- Sent from Reviewable.io -->

floreks commented:
Review status: 8 of 21 files reviewed at latest revision, 26 unresolved discussions.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 61 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ50WmWa3RTnnsCYfC_:-KQKmv05sbfPAtotBhC7:b-baqfba) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L61)):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> Thank you for this! Amazing keyword, I will use it everywhere now :)
</details>
Do i remember correctly that we've decided not to use `for ... of ...` syntax as it is not supported by IE (#344)? cc @bryk

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

Maybe `for ... in ...` works on IE?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

PiotrDabkowski commented:
Review status: 8 of 21 files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 61 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ50WmWa3RTnnsCYfC_:-KQKpboYumUdL_U-iarG:b3a1f82) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L61)):*
<details><summary><i>Previously, floreks (Sebastian Florek) wrote…</i></summary>
> Do i remember correctly that we've decided not to use `for ... of ...` syntax as it is not supported by IE (#344)? cc @bryk
> 
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
> 
> Maybe `for ... in ...` works on IE?
</details>
classes also don't work on IE but google closure compiler compiles it to javascript that is supported. I checked compiler app.js and I could not find for ... of anywhere so I think closure compiler compiles for ... of as well. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

floreks commented:
Review status: 8 of 21 files reviewed at latest revision, 26 unresolved discussions.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 61 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ50WmWa3RTnnsCYfC_:-KQKtUBBWh2Tg1tM6dDi:b-am1d52) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L61)):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> classes also don't work on IE but google closure compiler compiles it to javascript that is supported. I checked compiler app.js and I could not find for ... of anywhere so I think closure compiler compiles for ... of as well. 
</details>
Yes, It does. I don't remember what was the cause of changing `for ... of` to simple for syntax, but I know that we've been doing this :) Maybe @bryk remembers why.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

bryk commented:
*[src/app/frontend/common/components/graph/graph_component.js, line 61 \[r3\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQ50WmWa3RTnnsCYfC_:-KQL6qLO3SRQS_lk8vk4:b-99vuvj) ([raw file](https://github.com/kubernetes/dashboard/blob/d60fe3dc05fe451800f4a72ab6ecb7bd6fab550b/src/app/frontend/common/components/graph/graph_component.js#L61)):*
<details><summary><i>Previously, floreks (Sebastian Florek) wrote…</i></summary>
> Yes, It does. I don't remember what was the cause of changing `for ... of` to simple for syntax, but I know that we've been doing this :) Maybe @bryk remembers why.
</details>
It didn't work when we used JSCompiler for development mode. Now we use webpack and all should just work. Thanks for caring! :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

bryk commented:
One minor comment. I'll do final testing now :) 

---

Reviewed 1 of 16 files at r1, 3 of 5 files at r4, 8 of 10 files at r5, 1 of 2 files at r6.
Review status: 20 of 21 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/app/frontend/common/components/graph/graph_tick_formatters.js, line 44 \[r6\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQL8fCe1Ty8rXmzqn6D:-KQL8fCfVCGwW020MdA_:b-4vryez) ([raw file](https://github.com/kubernetes/dashboard/blob/f1ac168dbac2f2b5ba44787e1ea9dc33327dc46c/src/app/frontend/common/components/graph/graph_tick_formatters.js#L44)):*
> ```JavaScript
>  */
> export function formatMemoryUsage(d) {
>   return d === null ? 'N/A' : memoryFilter(precisionFilter)(d);
> ```

Can you move this to an i18n-ed block? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQL8vm20vDvLJgVkTDZ:byhajdf)*
<!-- Sent from Reviewable.io -->

olekzabl commented:
Reviewed 1 of 5 files at r4, 8 of 10 files at r5.
Review status: 20 of 21 files reviewed at latest revision, 9 unresolved discussions.

---

*[i18n/messages-en.xtb, line 586 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQKn3o1Q_vf9iiAkGjk:-KQKn3o1Q_vf9iiAkGjl:bqudcoc) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/i18n/messages-en.xtb#L586)):*
> ```
> 	<translation id="339987725992094148" key="MSG_GRAPH_CPU_AXIS_LABEL" source="/usr/local/google/home/pdabkowski/WebstormProjects/src/github.com/kubernetes/dashboard/.tmp/serve/app-dev.js" desc="Name of Y axis showing CPU usage.">CPU (cores)</translation>
> 	<translation id="2029236165141474072" key="MSG_GRAPH_MEMORY_AXIS_LABEL" source="/usr/local/google/home/pdabkowski/WebstormProjects/src/github.com/kubernetes/dashboard/.tmp/serve/app-dev.js" desc="Name of Y axis showing memory usage.">Memory (bytes)</translation>
> 	<translation id="4576303150557908117" key="MSG_GRAPH_CPU_LIMIT_LEGEND_LABEL" source="/usr/local/google/home/pdabkowski/WebstormProjects/src/github.com/kubernetes/dashboard/.tmp/serve/app-dev.js" desc="Name of the CPU usage metric as displayed in the legend.">CPU Limit</translation>
> ```

"usage" -> "limit"? (in the description)

---

*[i18n/messages-ja.xtb, line 775 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQKnEPzSbs8Vyv_dBhB:-KQKnEPzSbs8Vyv_dBhC:bh1k6zm) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/i18n/messages-ja.xtb#L775)):*
> ```
> 	<translation id="339987725992094148" key="MSG_GRAPH_CPU_AXIS_LABEL" source="/usr/local/google/home/pdabkowski/WebstormProjects/src/github.com/kubernetes/dashboard/.tmp/serve/app-dev.js" desc="Name of Y axis showing CPU usage.">CPU (cores)</translation>
> 	<translation id="2029236165141474072" key="MSG_GRAPH_MEMORY_AXIS_LABEL" source="/usr/local/google/home/pdabkowski/WebstormProjects/src/github.com/kubernetes/dashboard/.tmp/serve/app-dev.js" desc="Name of Y axis showing memory usage.">Memory (bytes)</translation>
> 	<translation id="4576303150557908117" key="MSG_GRAPH_CPU_LIMIT_LEGEND_LABEL" source="/usr/local/google/home/pdabkowski/WebstormProjects/src/github.com/kubernetes/dashboard/.tmp/serve/app-dev.js" desc="Name of the CPU usage metric as displayed in the legend.">CPU Limit</translation>
> ```

same as in ...-en.xtb

---

*[src/app/frontend/common/components/graph/graph_component.js, line 79 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQL4W2cbc2c8rWTqjRu:-KQL4W2cbc2c8rWTqjRv:b-ozo18i) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/src/app/frontend/common/components/graph/graph_component.js#L79)):*
> ```JavaScript
>                   'Can\'t display requested data - metrics have inconsistent types of y1 axis!');
>             }
>             y1max = Math.max(y1max, Math.max(...metric.dataPoints.map((e) => e.y)));
> ```

Once you finally determine y1max, you could increase it to prevent the graph from hitting the top boundary of the chart area.
(E.g. increase it by 10%?)
Of course it's subjective whether this would be nicer, but you probably remember that @bryk also mentioned this.

---

*[src/app/frontend/common/components/graph/graph_settings.js, line 86 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQLBMJprcc8zrUHKq-R:-KQLBMJqx53Wg3QXi4VR:b-etolys) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/src/app/frontend/common/components/graph/graph_settings.js#L86)):*
> ```JavaScript
>   MemoryAxisType: {
>     formatter: formatMemoryUsage,
>     label: i18n.MSG_GRAPH_CPU_AXIS_LABEL,
> ```

CPU -> MEMORY?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

PiotrDabkowski commented:
@bryk Done, PTAL

---

Review status: 18 of 21 files reviewed at latest revision, 9 unresolved discussions.

---

*[i18n/messages-en.xtb, line 586 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQKn3o1Q_vf9iiAkGjk:-KQLEtthXKcoVTQgsVY9:buosnfc) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/i18n/messages-en.xtb#L586)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> "usage" -> "limit"? (in the description)
</details>
Done, changed it in the source code.

---

*[i18n/messages-ja.xtb, line 775 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQKnEPzSbs8Vyv_dBhB:-KQLF1iOa9uj3m6EfKAd:buosnfc) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/i18n/messages-ja.xtb#L775)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> same as in ...-en.xtb
</details>
Done, changed it in the source code.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 79 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQL4W2cbc2c8rWTqjRu:-KQLFHdUT8uvtin_4rA_:b-9xjtta) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/src/app/frontend/common/components/graph/graph_component.js#L79)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> Once you finally determine y1max, you could increase it to prevent the graph from hitting the top boundary of the chart area.
> (E.g. increase it by 10%?)
> Of course it's subjective whether this would be nicer, but you probably remember that @bryk also mentioned this.
</details>
Thank you, it may actually look better with 10% margin. @bryk should I change it?

---

*[src/app/frontend/common/components/graph/graph_settings.js, line 86 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQLBMJprcc8zrUHKq-R:-KQLF_BF1e97eEWMWHsf:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/src/app/frontend/common/components/graph/graph_settings.js#L86)):*
<details><summary><i>Previously, olekzabl wrote…</i></summary>
> CPU -> MEMORY?
</details>
Done.

---

*[src/app/frontend/common/components/graph/graph_tick_formatters.js, line 44 \[r6\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQL8fCe1Ty8rXmzqn6D:-KQLFaB5ob6XqYqGUwSd:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/f1ac168dbac2f2b5ba44787e1ea9dc33327dc46c/src/app/frontend/common/components/graph/graph_tick_formatters.js#L44)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you move this to an i18n-ed block? 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLFbJ2EDEwsqFxV1FR:bubllvs)*
<!-- Sent from Reviewable.io -->

bryk commented:
This does not work on `gulp serve:prod`. Can you work with Olek to debug this? He knows how do do this :) 

---

Review status: 16 of 21 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLN6t7Gxc_CTIkmfCX:b-rfep67)*
<!-- Sent from Reviewable.io -->

bryk commented:
By quick look i think this:
```js
export const axisSettings = {
  CPUAxisType: {
    formatter: formatCpuUsage,
    label: i18n.MSG_GRAPH_CPU_AXIS_LABEL,
  },
  MemoryAxisType: {
    formatter: formatMemoryUsage,
    label: i18n.MSG_GRAPH_MEMORY_AXIS_LABEL,
  },
  TimeAxisType: {
    formatter: formatTime,
    label: i18n.MSG_GRAPH_TIME_AXIS_LABEL,
  },
};

```

And the other setting file.

Basically, all keys need to be in '`'. Otherwise compiler will rename them 

---

Review status: 16 of 21 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLNnAroIZR9Ps5JoZV:btn1uyo)*
<!-- Sent from Reviewable.io -->

bryk commented:
![Screen Shot 2016-08-29 at 3.39.19 PM.png](https://files.reviewableusercontent.io/images/reviewable/1159999/hyQxf6HLSlI1s4tLTP1USwD3/Screen%20Shot%202016-08-29%20at%203.39.19%20PM.png)

Sparklines on service pages are broken. Fix? 

---

Review status: 16 of 21 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLWvmcIE_OwCxvW4-5:b-k5t2mu)*
<!-- Sent from Reviewable.io -->

olekzabl commented:
Reviewed 2 of 5 files at r4, 1 of 10 files at r5, 1 of 2 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142)*
<!-- Sent from Reviewable.io -->

PiotrDabkowski commented:
Apologies, I did not notice another pair of sparklines. Fixed.

---

Review status: 17 of 21 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLaogSj8bHXlE54KzN:b-drrptd)*
<!-- Sent from Reviewable.io -->

bryk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Let's wait for Olek and Travis and merge :) 

---

Reviewed 1 of 10 files at r3, 2 of 5 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/app/frontend/common/components/graph/graph_component.js, line 79 \[r5\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-KQL4W2cbc2c8rWTqjRu:-KQLakB_NZkhDqhtVlCZ:bm5py0o) ([raw file](https://github.com/kubernetes/dashboard/blob/58fc7b01d499cb7049a80027a075f4487c688dee/src/app/frontend/common/components/graph/graph_component.js#L79)):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> Thank you, it may actually look better with 10% margin. @bryk should I change it?
</details>
Yeah, should be fine. But how about doing look&feel changes in a next PR? The graphs need design polish anyway :) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLb1M26emYkWx6PHLQ:bsb8muj)*
<!-- Sent from Reviewable.io -->

olekzabl commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLbU2QiYw4OQlHwjPm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bryk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Will merge when travis is happy 

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1142#-:-KQLe-Rh4VCXaz3VMNbc:b-q04fv2)*
<!-- Sent from Reviewable.io -->
